### PR TITLE
support azure assistant api

### PR DESCRIPTION
**Pull request by IANTHEREAL**
## Why are these changes needed?

while using azure gpt assistant api, `model` in `config_list` would cause error:  `NotFoundError: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}`. 

Following the azure tutorial - https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/assistant#create-an-assistant, we should remove `model`  from `config_list` 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by IANTHEREAL** [comment]
@sonichi  @gagb  Hi, how should I add a azure assistant api testing? It requires an azure openai account.

File: autogen/agentchat/contrib/gpt_assistant_agent.py [hunk 1]
```diff
@@ -53,9 +53,14 @@ def __init__(
                 - Other kwargs: Except verbose, others are passed directly to ConversableAgent.
         """
         # Use AutoGen OpenAIWrapper to create a client
-        oai_wrapper = OpenAIWrapper(**llm_config)
-        if len(oai_wrapper._clients) > 1:
+        model_name = "gpt-4-1106-preview"
```
Comment (sonichi) [inline autogen/agentchat/contrib/gpt_assistant_agent.py:56]: What about requiring model_name to be provided by the user and not preset here? For other agents we require it too.
It's OK to do it in a separate PR.

File: autogen/agentchat/contrib/gpt_assistant_agent.py [hunk 2]
```diff
@@ -53,9 +53,14 @@ def __init__(
...
+        if len(llm_config["config_list"]) > 1:
             logger.warning("GPT Assistant only supports one OpenAI client. Using the first client in the list.")
+
+        if len(llm_config["config_list"]) > 0:
+            model_name = llm_config["config_list"][0].pop("model", "gpt-4-1106-preview")
```
Comment (sonichi) [inline autogen/agentchat/contrib/gpt_assistant_agent.py:61]: The separation of `model` from other constructors is already handled by `OpenAIWrapper` in its constructor.
It's OK for this PR.

File: autogen/agentchat/contrib/gpt_assistant_agent.py [hunk 3]
```diff
@@ -79,7 +84,7 @@ def __init__(
                     name=name,
                     instructions=instructions,
                     tools=llm_config.get("tools", []),
-                    model=llm_config.get("model", "gpt-4-1106-preview"),
+                    model=model_name,
                     file_ids=llm_config.get("file_ids", []),
```
Comment (sonichi) [inline autogen/agentchat/contrib/gpt_assistant_agent.py:88]: A cleaner way is to make everything about `create` handled by `OpenAIWrapper` and not here.
It's OK to do that in a separate PR.

**Review by sonichi — COMMENTED** [review]
I made a few suggestions for future PR.
cc @gagb @olgavrou @ekzhu @BeibinLi @randombet @kinfey 

For testing on GitHub, you can use `config_list_from_json` and add `"api_type": ["2024-02-15-preview"]` into the `filter_dict`.

File: autogen/agentchat/contrib/gpt_assistant_agent.py (hunk 1 above)
Comment (IANTHEREAL) [reply autogen/agentchat/contrib/gpt_assistant_agent.py:56]: I agree with adopting a unified approach to handle this. I'm not particularly familiar with the logic of OpenAIWrapper, but I can look into it later.

File: autogen/agentchat/contrib/gpt_assistant_agent.py (hunk 2 above)
Comment (IANTHEREAL) [reply autogen/agentchat/contrib/gpt_assistant_agent.py:61]: Ditto. We can start with a workaround for now, and I'll look into a more unified approach later on.

**Comment by IANTHEREAL** [comment]
> For testing on GitHub, you can use config_list_from_json and add "api_type": ["2024-02-15-preview"] into the filter_dict.

Ok. I'll see how to run the same test code twice using both OpenAI and Azure API separately.

**Comment by IANTHEREAL** [comment]
@sonichi I'm not quite sure about the best way to add Azure tests, could you take a look at my latest commit and see if you have any suggestions?

File: test/agentchat/contrib/test_gpt_assistant.py [hunk 4]
```diff
@@ -24,7 +24,7 @@
 
 if not skip:
     config_list = autogen.config_list_from_json(
-        OAI_CONFIG_LIST, file_location=KEY_LOC, filter_dict={"api_type": ["openai"]}
+        OAI_CONFIG_LIST, file_location=KEY_LOC, filter_dict={"api_type": ["openai", "azure"]}
     )
```
Comment (sonichi) [inline test/agentchat/contrib/test_gpt_assistant.py:28]: Keep the original config_list as openai_config_list
Add an aoai_config_list with filter_dict={"api_version": ["2024-02-15-preview"]}
Then test these two config_lists separately

**Review by archit15singh — APPROVED** [review]

File: autogen/agentchat/contrib/gpt_assistant_agent.py [hunk 5]
```diff
@@ -53,9 +53,14 @@ def __init__(
                 - Other kwargs: Except verbose, others are passed directly to ConversableAgent.
         """
         # Use AutoGen OpenAIWrapper to create a client
-        oai_wrapper = OpenAIWrapper(**llm_config)
+        openai_client_cfg = None
+        if llm_config and llm_config.get("config_list") is not None and len(llm_config["config_list"]) > 0:
+            openai_client_cfg = llm_config["config_list"][0].copy()
+
+        oai_wrapper = OpenAIWrapper(base_config=openai_client_cfg)
```
Comment (sonichi) [inline autogen/agentchat/contrib/gpt_assistant_agent.py:60]:
```suggestion
        oai_wrapper = OpenAIWrapper(**llm_config)
        llm_config = oai_wrapper._config_list[0]
```
We should not access private methods here, so my suggestion needs to be later modified when we add a public property.

File: test/agentchat/contrib/test_gpt_assistant.py [hunk 6]
```diff
@@ -23,24 +23,35 @@
...
+    assert len(openai_config_list) > 0
+    assert len(aoai_config_list) > 0
 
 
 @pytest.mark.skipif(
     sys.platform in ["darwin", "win32"] or skip,
     reason="do not run on MacOS or windows OR dependency is not installed OR requested to skip",
 )
 def test_gpt_assistant_chat() -> None:
+    for gpt_config in [openai_config_list, aoai_config_list, None]:
```
Comment (sonichi) [inline test/agentchat/contrib/test_gpt_assistant.py:50]: None would fail I think.

**Comment by archit15singh** [comment]
🚀

**Comment by gagb** [comment]
@IANTHEREAL when I run `notebook/agentchat_oai_assistant_twoagents_basic.ipynb` on this branch with my OpenAI keys, I get:
```console
OpenAIError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable
````
I do not get this error on the main.

**Comment by IANTHEREAL** [comment]
@gagb mine is ok, did you use my latest commits?

image

File: autogen/agentchat/contrib/gpt_assistant_agent.py (hunk 5 above)
Comment (IANTHEREAL) [reply autogen/agentchat/contrib/gpt_assistant_agent.py:60]: The modifications you suggested will lead to this error: NotFoundError: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}

File: test/agentchat/contrib/test_gpt_assistant.py (hunk 6 above)
Comment (IANTHEREAL) [reply test/agentchat/contrib/test_gpt_assistant.py:50]: It's ok in my local testing

**Comment by gagb** [comment]
> @gagb mine is ok, did you use my latest commits?
> 
> image

I am on the latest commit. But still getting the error:

image

**Comment by IANTHEREAL** [comment]
@gagb I only set OPENAI_API_KEY in the environment variables, so it does not show up in the OAI_CONFIG_LIST configuration.

File: test/agentchat/contrib/test_gpt_assistant.py (hunk 6 above)
Comment (IANTHEREAL) [reply test/agentchat/contrib/test_gpt_assistant.py]: fixed

**Comment by gagb** [comment]
> @gagb I only set OPENAI_API_KEY in the environment variables, so it does not show up in the OAI_CONFIG_LIST configuration.

The code should work if key is specified in the config list instead of openai_api_key env variable? That's where its currently failing for me.

**Comment by IANTHEREAL** [comment]
@gagb can you give it a try? I just fixed the code, I got the parameters of OpenAIWrapper wrong

**Review by sonichi — APPROVED** [review]
There is some problem. When users provide the model in the `llm_config`, it will be ignored. 
Prior to this PR, when the model is provided in `config_list`, it will be ignored. Neither is comprehensive.
We can fix this issue later.

**Comment by gagb** [comment]
> @gagb can you give it a try? I just fixed the code, I got the parameters of OpenAIWrapper wrong

Btw I did test it today. It worked with openai assistants, but I got stuck at trying to find an azure key that supported assistant API. The one I had access to did not :( I am working on getting it
